lib/datatype/data-type-new.js

```diff
@@ -0,0 +1,206 @@
+ */
+/**
+ * @typedef {import('../types.js').MapeoDocMap} MapeoDocMap
+ */
+/**
+ * @typedef {import('../types.js').MapeoValueMap} MapeoValueMap
+ */
+/**
+ * @typedef {`${MapeoDoc['schemaName']}Table`} MapeoDocTableName
+ */
+/**
+ * @template T
+ * @typedef {T[(keyof T) & MapeoDocTableName]} GetMapeoDocTables
+ */
+/**
+ * Union of Drizzle schema tables that correspond to MapeoDoc types (e.g. excluding backlink tables and other utility tables)
+ * @typedef {GetMapeoDocTables<import('../schema/project.js')> | GetMapeoDocTables<import('../schema/client.js')>} MapeoDocTables
+ */
+/**
+ * @typedef {{ [K in MapeoDocTables['_']['name']]: Extract<MapeoDocTables, { _: { name: K }}> }} MapeoDocTablesMap
+ */
+/**
+ * @template T
+ * @template {keyof any} K
+ * @typedef {T extends any ? Omit<T, K> : never} OmitUnion
+ */
+
+function generateId() {
+  return randomBytes(32).toString('hex')
+}
+function generateDate() {
+  return new Date().toISOString()
+}
+
+/**
+ * @template {import('../datastore/new-data-store.js').DataStore} TDataStore
+ * @template {TDataStore['schemas'][number]} TSchemaName
+ * @template {MapeoDocTablesMap[TSchemaName]} TTable
+ * @template {MapeoDocMap[TSchemaName]} TDoc
+ * @template {MapeoValueMap[TSchemaName]} TValue
+ */
+export class DataType {
+  #dataStore
+  #table
+  #getPermissions
+  #schemaName
+  #sql
+
+  /**
+   *
+   * @param {object} opts
+   * @param {TTable} opts.table
+   * @param {TDataStore} opts.dataStore
+   * @param {import('drizzle-orm/better-sqlite3').BetterSQLite3Database} opts.db
+   * @param {() => any} [opts.getPermissions]
+   */
+  constructor({ dataStore, table, getPermissions, db }) {
+    this.#dataStore = dataStore
+    this.#table = table
+    this.#schemaName = /** @type {TSchemaName} */ (getTableConfig(table).name)
+    this.#getPermissions = getPermissions
+    this.#sql = {
+      getByDocId: db
+        .select()
+        .from(table)
+        .where(eq(table.docId, placeholder('docId')))
+        .prepare(),
+      getMany: db.select().from(table).prepare(),
+    }
+  }
+
+  /**
+   * @template {import('type-fest').Exact<TValue, T>} T
+   * @param {T} value
+   */
+  async create(value) {
+    if (!validate(this.#schemaName, value)) {
+      // TODO: pass through errors from validate functions
+      throw new Error('Invalid value ' + value)
+    }
+    const nowDateString = generateDate()
+    /** @type {OmitUnion<MapeoDoc, 'versionId'>} */
+    const doc = {
+      ...value,
+      docId: generateId(),
+      createdAt: nowDateString,
+      updatedAt: nowDateString,
+      links: [],
+    }
+
+    // TS can't track the relationship between TDoc and TValue, so doc above is
+    // typed as MapeoDoc (without versionId) rather than as TDoc.
+    await this.#dataStore.write(/** @type {TDoc} */ (doc))
+    return this.getByDocId(doc.docId)
+  }
+
+  /**
+   * @param {string} docId
+   */
+  async getByDocId(docId) {
+    return deNullify(this.#sql.getByDocId.get({ docId }))
+  }
```

Comment:
I made all these methods async to allow us to switch to running sql queries in a separate thread if we need to in the future, or if we need to switch out better-sqlite3 for a different sqlite wrapper that is async. Everything needs to be async from the client-side anyway, so I think it's ok.

Reply:
fair enough! was thinking that it would be more flexible to keep as sync because the caller could make it async if desired (sync -> async is easier than async -> sync)

also curious given the first note of this section in the better-sqlite3 docs, which I found interesting:

https://github.com/WiseLibs/better-sqlite3#why-should-i-use-this-instead-of-node-sqlite3

> node-sqlite3 uses asynchronous APIs for tasks that are either CPU-bound or serialized. That's not only bad design, but it wastes tons of resources. It also causes mutex thrashing which has devastating effects on performance.

not sure if it's referring to the public-facing APIs or stuff it does internally, so maybe it doesn't apply here